sqlx-core/src/pool/inner.rs

```diff
@@ -291,7 +291,13 @@ impl<DB: Database> PoolInner<DB> {
                     };
 
                     // Attempt to connect...
-                    return self.connect(deadline, guard).await;
+                    let pool = self.clone();
+                    return crate::rt::spawn(async move {
+                        crate::rt::timeout(pool.options.connect_timeout, async move {
```

Comment:
There's already a timeout inside `connect()`, you just need to convert `connect_timeout` to an `Instant` and pass it as `deadline`.